src/dos/dos_mscdex.cpp

```diff
@@ -400,23 +383,20 @@ int CMscdex::AddDrive(uint16_t _drive, const char* physicalPath, uint8_t& subUni
 	DOS_DeviceHeader devHeader(PhysicalMake(rootDriverHeaderSeg,0));
 	devHeader.SetNumSubUnits(devHeader.GetNumSubUnits()+1);
 
-	if (dinfo[0].drive-1==_drive) {
-		CDROM_Interface *_cdrom = cdrom[numDrives];
-		CDROM_Interface_Image *_cdimg = CDROM_Interface_Image::images[numDrives];
-		for (uint16_t i=GetNumDrives(); i>0; i--) {
-			dinfo[i] = dinfo[i-1];
-			cdrom[i] = cdrom[i-1];
-			CDROM_Interface_Image::images[i] = CDROM_Interface_Image::images[i-1];
+	if (dinfo[0].drive - 1 == _drive) {
+		auto _cdrom = std::move(CDROM::cdroms[numDrives]);
+		for (uint16_t i = GetNumDrives(); i > 0; i--) {
+			dinfo[i]         = dinfo[i - 1];
+			CDROM::cdroms[i] = std::move(CDROM::cdroms[i - 1]);
 		}
-		cdrom[0] = _cdrom;
-		CDROM_Interface_Image::images[0] = _cdimg;
-		dinfo[0].drive		= (uint8_t)_drive;
-		dinfo[0].physDrive	= (uint8_t)toupper(physicalPath[0]);
-		subUnit = 0;
+		CDROM::cdroms[0]   = std::move(_cdrom);
+		dinfo[0].drive     = (uint8_t)_drive;
+		dinfo[0].physDrive = (uint8_t)toupper(physicalPath[0]);
+		subUnit            = 0;
```

Comment:
Can you double-check the logic here any maybe clean this up a bit?  There's lot of `std::move` both in and out of the array and I'm not confident that this is doing the same thing that the old code was.

It would be better if you could keep the smart pointer as a local variable and then do `std::move` once to move it into the array.

Reply:
The loop was already there. The only difference between the old and the new is that we only have to rotate things in two places (CDROM::cdroms and dinfo) instead of three (dinfo, cdrom and CDROM_Interface_Image::images).

All this code performs a push_front in an array. I'll add a comment there and see if I can simplify this. CDROM::cdroms should actually be an std::vector...